main/src/ca/uwaterloo/flix/language/phase/unification/BoolFormula.scala

```diff
@@ -230,6 +230,50 @@ object BoolFormula {
 
     override def isFalse(f: BoolFormula): Boolean = f == BoolFormula.False
 
+    override def isVar(f: BoolFormula): Boolean = f match {
+      case Var(_) => true
+      case _ => false
+    }
+
+    override def satisfiable(f: BoolFormula): Boolean = f match {
+      case BoolFormula.True => true
+      case BoolFormula.False => false
+      case BoolFormula.Var(_) => true
+      case _ =>
+        enumerateAssignmentsLazy(freeVars(f)).exists(a => checkAssignment(f, a))
+    }
+
+    /**
+      * Enumerates all assignments to the boolean variables in `vars` in a lazy manner.
+      */
+    private def enumerateAssignmentsLazy(vars: SortedSet[Int]): LazyList[Map[Int, Boolean]] = {
+      LazyList.from(0)
+              .map(n => padBinaryStringToLength(n.toBinaryString, vars.size)
+                         .toCharArray
+                         .zip(vars)
+                         .foldLeft(Map.empty[Int,Boolean])((m,ci) => m ++ Map(ci._2 -> (ci._1 == '1'))))
+              .take(scala.math.pow(2,vars.size).toInt)
+    }
+
+    /**
+      * Adds the prefix "0" until the string `s` has length `n`
+      */
+    private def padBinaryStringToLength(s: String, n: Int): String = {
```

Comment:
Oh god. What have I asked for? 😮

Reply:
I think the enumeration separate from the evaluation is the problem. How about this structure:

```
def foo(f: Formula, l: List[Int], env: Map[Int, Boolean]): Boolean = l {
  case Nil => // all variables bound: 
                checkAssignment(f, env)
  case x :: xs => // two recursive calls, note the use of || which is lazy/short circuit.
    foo(f, xs, env + (x -> true)) || foo(f, xs, env + (x -> false))
```

(Note the env can be replaced by a set that only stores the true variables.)